CopyKeys/CopyKeys.ps1

```diff
@@ -1,1468 +1,1676 @@
+    Compare-Permissions -KeyVaultName $KeyVaultResource.Name -PermissionsRequired $SourcePermissions `
+        -AccessPolicies $KeyVaultResource.Properties.AccessPolicies -ResourceObject
+
+    return $KeyVaultResource
+}
+
+### <summary>
+### Create a secret in the target key vault.
+### </summary>
+### <param name="Secret">Value of the secret text.</param>
+### <param name="ContentType">Type of secret to be created - Wrapped BEK or BEK.</param>
+function Create-Secret(
+    $Secret,
+    [string]$ContentType)
+{
+    $SecureSecret = ConvertTo-SecureString $Secret -AsPlainText -Force
+    $OutputSecret = Set-AzKeyVaultSecret -VaultName $TargetBekVault -Name $BekSecret.Name -SecretValue `
+        $SecureSecret -tags $BekTags -ContentType $ContentType
+
+    $OutputLogger.Log(
+        $MyInvocation,
+        "Copying 'Disk Encryption Key' for '$VmName'.",
+        [LogType]::OUTPUT)
+    $OutputLogger.Log(
+        $MyInvocation,
+        "TargetBEKVault: $TargetBekVault",
+        [LogType]::OUTPUT)
+    $OutputLogger.Log(
+        $MyInvocation,
+        "TargetBEKId: $($OutputSecret.Id)",
+        [LogType]::OUTPUT)
+}
+
+### <summary>
+### Main flow of code for copying keys.
+### </summary>
+### <return name="CompletedList">List of VMs for which CopyKeys ran successfully</return>
+function Start-CopyKeys
+{
+    $Context = Get-AzContext
+
+    if($null -eq $Context)
+    {
+        $SuppressOutput = Login-AzAccount -ErrorAction Stop
+    }
+
+    $CompletedList = @()
+    $UserInputs = New-Object System.Collections.Hashtable
+    Write-Verbose "Starting user interface to get inputs"
+    Generate-UserInterface
+
+    $ResourceGroupName = $UserInputs["ResourceGroupName"]
+    $VmNameArray = $UserInputs["VmNameArray"]
+    $TargetLocation = $UserInputs["TargetLocation"]
+    $TargetBekVault = $UserInputs["TargetBekVault"]
+    $TargetKekVault = $UserInputs["TargetKekVault"]
+
+
+    if($null -eq $Context)
+    {
+        $Context = Get-AzContext
+    }
+
+    $OutputLogger.Log(
+        $MyInvocation,
+        "SubscriptionId: $($Context.Subscription.Id)",
+        [LogType]::INFO)
+    $OutputLogger.LogObject(
+        $MyInvocation,
+        $UserInputs,
+        "User inputs.",
+        [LogType]::INFO)
+
+    $TenantId = $Context.Tenant.Id
+    $AccessToken = (Get-AzAccessToken -ResourceTypeName KeyVault).Token
+    $UserPrincipalName = (Get-AzContext).Account.Id
+    $UserId = (Get-AzAdUser -UserPrincipalName $UserPrincipalName).Id
+
+    $OutputLogger.Log(
+        $MyInvocation,
+        "`nStarting CopyKeys for UserId: $UserId, UserPrincipalName: $UserPrincipalName",
+        [LogType]::OUTPUT)
+
+    $IsFirstBekVault = $IsFirstKekVault = $true
+    $FirstBekVault = $FirstKekVault = $null
+    $IsBekKeyVaultNew = $IsKekKeyVaultNew = $false
+
+    $SourceList = New-Sources -VmArray $VmNameArray -SourceResourceGroupName $ResourceGroupName
+
+    foreach($Source in $SourceList)
```

Comment:
Can we add an info ballon in the ui..........which tells user, why the vm is not showing up in the drop down

Reply:
We are already doing this. It's not exactly an info balloon but in case of 0 VMs we do put a message saying that the selected RG doesnt have any encrypted VMs